### PR TITLE
spec(rest_api) - proposed spec of rest api

### PR DESCRIPTION
**Pull request by sedouard**
This spec proposes a json api rc3 complaint rest api for the connected car
backend.

### CODE REVIEW

**Comment by sedouard** [comment]
@dougseven included your changes from my fork (you can just add in-line comments if its nothing big).

Using the `includes` top-level key you can add any additional data about the requested entity. In our case `/vehicle:id` can either include a `linkage` object specifying an array of `type` and `id` keys which point to the models placed in `includes` OR you can simply just point the `trips` key to the url which will return the related models using `related`.

Take a peek. I have a set of ember unit tests on the client consuming the api currently as written. So once we lock it down we can start implementing client-side stuff independent from the backend and hook it up together when you're ready to go live.

**Comment by felixrieseberg** [comment]
Looks pretty good to me - we can serialize this for ember data by just doing some normalization, which seems easy enough. The relations might run deep, but we can probably async that away.

**Comment by sedouard** [comment]
I got everything working client side with the json api adapter for ember
(it extends the normal rest adapter). Its handling all the normalization
for us.

**Comment by felixrieseberg** [comment]
Oh, awesome then.

**Comment by sedouard** [comment]
Good amount of changes here on the client side...

@dougseven double check at specs/rest-api.md. Only major change here is that `location` has to have a unique `id` key, otherwise our parsers will choke. 

These changes add models: `driver`, `event`, `location`, `trip`, `vehicle` and associated unit tests. Mocking is implemented at the jquery ajax level via [jquery-mockjax](https://github.com/jakerella/jquery-mockjax). Loading of this module is done through `Brocfile.js` depending on if we're in test mode. 

@felixrieseberg I had to do a good amount of digging to understand how ember cli sets up it's default resolver behavior which is why I had to make a decent amount of code change there. Basically I wrote a custom resolver since we have 2 types of resolvers and serializers and different models utilize these.

Tests all pass in isolation, which allows us to start consuming apis today in the dashboard until actual implementation happens on the backend.

If all looks good to you guys I'll merge and we can move forward.

**Comment by felixrieseberg** [comment]
That looks decent, but let's do some iterations here. I am for instance not sure that API mocking should be done by the client (ember cli comes with http mock); I might also have some  ideas around adapters and serializers to avoid having that "surprise" code in app.js ;)

**Comment by felixrieseberg** [comment]
Elaborating a bit on this thing above: I'd like to ensure that we can easily give it to somebody else - ideally without having to explain _anything_. This is a good solution, but to that end, we might want to stick closer to the Ember ways. Here are three immediate thoughts I had:

1) I'd expect serializers and adapters to be defined in the respective folders. I'm totally fine with having a simple import/export file for each model - it's verbose, but we wouldn't have to explain it.
2) It's becoming kind of obvious that Nitrogen won't serve any purpose besides managing messages. If that's the case, we might move most of it's logic into the vehicle model (and get rid of all the auxiliary code). This would also leave us with only one adapter and serializer, making point 1) entirely redundant (wouldn't that be nice :grinning:) 
3) I'm not sure that I want jQuery to mock our API calls. Could we go with http mock? We'd still use 99% of the work here, which is :+1: 

In summary: Let's wait until @timpark is back so that we can tie down the API. If we can, I'd like to go with the JSON API for the app, moving the messaging part into the vehicle model, making the resolver customization (ideally) redundant, because we wouldn't need LSA anymore. Thoughts?

**Comment by sedouard** [comment]
I agree that the stuff going on in `app.js` is sort of a circus. This is due to the fact that default ember cli adapter resolving seems to first look for `ApplicationAdapter` and then searches for adapters by the respective model name. Irregardless of the fact that you may set `App.VehicleAdapter` to JsonApiAdapter. This works in the Azure Storage explorer project because we have 2 adapters with the same name as the models.

Here's responses to each of your bullets:

1) We can move the `.extends` portion to the respective serializer/adapter files and export those. This would keep the full definition of the serializers & adapters in 1 file, the resolver can be worked into its own file as well.
2) YES I completely agree, I found the code like this so I'm not sure if things were setup with nitrogen like that for a reason. This would completely make 1) moot :-). However this seems like another work item which I'll track as https://github.com/felixrieseberg/OxideFleet/issues/57
3) Here's the stack trace from a failed XHR from ember's rest api adapter:

Error
    at Error.ServerError (json-api-adapter.js:186)
    at DS.JsonApiAdapter.DS.RESTAdapter.extend.ajaxError (json-api-adapter.js:168)
    at apply (ember.debug.js:18040)
    at superWrapper [as ajaxError] (ember.debug.js:17594)
    at ember$data$lib$system$adapter$$Adapter.extend.ajax.Ember.RSVP.Promise.hash.error (rest-adapter.js:789)
    at jQuery.Callbacks.fire (jquery.js:3094)
    at Object.jQuery.Callbacks.self.fireWith [as rejectWith] (jquery.js:3206)
    at done (jquery.js:8261)
    at XMLHttpRequest.jQuery.ajaxTransport.send.callback (jquery.js:8594)

If ember is using jQuery for XHR doesn't it make sense to use `mockjax`? That was my reasoning...

Sure makes sense to hold off for timpark. I wanted to get the ball rolling on this to unblock. I made the PR based of the initial schema doug sent for a vehicle and made it fit into json api spec. 

The proposed draft spec does conform to the standards but maybe theres data or something that he wouldn't agree with. Either way this is a good setup as we've already plugged in an adapter that understands jsonapi as well as models and mocking infrastructure.

I'll leave this open until we fix https://github.com/felixrieseberg/OxideFleet/issues/57 and @timfpark gives feedback on the api spec.

**Comment by felixrieseberg** [comment]
I like the implementation (it's pretty good), I just have a few ideas to improve it. I like to keep to the "official ember way" so that we can just point at the documentation without having to explain anything. The changes wouldn't even be that big (and we'll end up with less code, I promise!)

 * Move the `if/else` for the JSON API host into the config (`ENV.jsonHost` or something like that), importing the host variable in the adapter files.
 * Instead of defining the adapters in `app.js`, have three files in `adapters`. `application` defines the JSON API adapter, while `device` and `user` define the LSA adapter.
* Instead of defining the serializers in `app.js`, have three files in `serializers`. `application` defines the JSON Serializer, while `device` and `user` define the LSA serializer.
* The point about Ember doing AJAX in jQuery is a good one. I think I can solve the documentation by making a small `readme` in a `server` folder in the root (where Ember cli usually keeps it mock apis). I can put that on my to-do list.

Everything else is :+1: anyway, so with those changes (if they pass tests), I'd give :shipit: for merging!

**Comment by sedouard** [comment]
Cool - I'm open to using whatever mocking api by the way I'm not in love with mockjax I just picked it because ember was doing query but didn't realize there was a preferred ember-cli way.

I'll make the changes listed above - but just curious... is there REALLY no way to reuse an adapter/serializer across multiple models? It just seems that really sucks.

**Comment by sedouard** [comment]
Looks much better now thanks @felixrieseberg for the ember-cli tips!
